modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogCompactionRunner.java

```diff
@@ -0,0 +1,343 @@
+    }
+
+    /** Called when the low watermark has been changed. */
+    public CompletableFuture<Boolean> onLowWatermarkChanged(HybridTimestamp newLowWatermark) {
+        lowWatermark = newLowWatermark;
+
+        triggerCompaction(newLowWatermark);
+
+        return CompletableFutures.falseCompletedFuture();
+    }
+
+    @TestOnly
+    CompletableFuture<Boolean> lastRunFuture() {
+        return lastRunFuture;
+    }
+
+    /** Starts the catalog compaction routine. */
+    CompletableFuture<Boolean> triggerCompaction(@Nullable HybridTimestamp lwm) {
+        if (lwm == null || !topologyService.localMember().name().equals(compactionCoordinatorNodeName)) {
+            return CompletableFutures.falseCompletedFuture();
+        }
+
+        return inBusyLock(busyLock, () -> {
+            CompletableFuture<Boolean> fut = lastRunFuture;
+
+            if (!fut.isDone()) {
+                LOG.info("Catalog compaction is already in progress, skipping (timestamp={})", lwm.longValue());
+
+                return CompletableFutures.falseCompletedFuture();
+            }
+
+            fut = startCompaction(logicalTopologyService.localLogicalTopology())
+                    .whenComplete((res, ex) -> {
+                        if (ex != null) {
+                            LOG.warn("Catalog compaction has failed (timestamp={})", ex, lwm.longValue());
+                        } else if (LOG.isDebugEnabled()) {
+                            if (res) {
+                                LOG.debug("Catalog compaction completed successfully (timestamp={})", lwm.longValue());
+                            } else {
+                                LOG.debug("Catalog compaction skipped (timestamp={})", lwm.longValue());
+                            }
+                        }
+                    });
+
+            lastRunFuture = fut;
+
+            return fut;
+        });
+    }
+
+    private CompletableFuture<Boolean> startCompaction(LogicalTopologySnapshot topologySnapshot) {
+        return determineGlobalMinimumRequiredTime(topologySnapshot.nodes())
+                .thenComposeAsync(ts -> {
+                    Catalog catalog = catalogByTsNullable(ts);
+
+                    if (catalog == null) {
+                        return CompletableFutures.falseCompletedFuture();
+                    }
+
+                    return requiredNodes(catalog)
+                            .thenCompose(requiredNodes -> {
+                                List<String> missingNodes = missingNodes(requiredNodes, topologySnapshot.nodes());
+
+                                if (!missingNodes.isEmpty()) {
+                                    if (LOG.isDebugEnabled()) {
+                                        LOG.debug("Catalog compaction aborted due to missing cluster members (nodes={})", missingNodes);
+                                    }
+
+                                    return CompletableFutures.falseCompletedFuture();
+                                }
```

Comment:
so, we do distributed operation just to discard it later because of local computations... I think, it's better to change the order. Also, let's check that there is what to compact (at least locally) before we send broadcast message

Reply:
> let's check that there is what to compact (at least locally) before we send broadcast message

I believe I did it